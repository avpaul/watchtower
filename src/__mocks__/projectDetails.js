export default {
  id: 1,
  name: 'test project',
  type: 'internal',
  technologies: [
    {
      name: 'react',
      label: 'react',
      id: 51952
    },
    {
      name: 'ruby',
      label: 'ruby',
      id: 53375
    }
  ],
  documents: [
    {
      name: 'WatchTower Product Brief',
      url:
        'https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf'
    },
    {
      name: 'WatchTower Product Brief 2',
      url:
        'https: //www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf'
    },
    {
      name: 'WatchTower Product Brief 3',
      url:
        'https: //www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf'
    }
  ],
  manager: {
    id: 1,
    email: 'philip@test.com',
    name: 'philip',
    picture:
      'https//secure.i.telegraph.co.uk/multimedia/archive/03249/archetypal-male-fa_3249635c.jpg',
    created_at: null,
    updated_at: null
  },
  channels: [{ id: 1, name: 'slack channel' }],
  mockups: 'http://google.com',
  tagline: null,
  about: 'This is the about of the test project',
  logo:
    'https: //cdn1.imggmi.com/uploads/2019/6/13/124459eb1455fd8f39aa700e3959e375-full.png',
  created_at: null,
  updated_at: null,
  engineers: [
    {
      id: 1,
      first_name: 'Chidozie',
      last_name: 'Nwoga',
      email: 'chidozie.nwoga@andela.com',
      picture:
        'https://media.haircutinspiration.com/photos/20181204023653/Back-and-Side-Square.jpg',
      sims_project: 'AuthorsHaven',
      sims_project_technology: 'Python/Django',
      sims_start_date: '2018-10-31 00:00:00',
      sims_end_date: '2019-01-10 00:00:00',
      sims_manager: 'Duncan Ireri',
      apprenticeship_project: 'WatchTower',
      apprenticeship_technology: 'Javascript/Ruby',
      apprenticeship_start_date: '2019-06-11 00:00:00',
      apprenticeship_end_date: '2019-06-26 00:00:00',
      apprenticeship_manager: 'Njenga Njunge',
      account_active: false,
      email_sent: false,
      project_id: 1,
      cohort: 'Lagos-43',
      project_role_name: 'Engineer',
      created_at: '2019-06-12 19:52:36',
      updated_at: '2019-06-12 19: 52: 36'
    },
    {
      id: 1,
      first_name: 'Monica',
      last_name: 'Jonzi',
      email: 'monica.jonzi@andela.com',
      picture:
        'https://media.haircutinspiration.com/photos/20181204023653/Back-and-Side-Square.jpg',
      sims_project: 'AuthorsHaven',
      sims_project_technology: 'Python/Django',
      sims_start_date: '2018-10-31 00:00:00',
      sims_end_date: '2019-01-10 00:00:00',
      sims_manager: 'Duncan Ireri',
      apprenticeship_project: 'WatchTower',
      apprenticeship_technology: 'Javascript/Ruby',
      apprenticeship_start_date: '2019-06-11 00:00:00',
      apprenticeship_end_date: '2019-06-26 00:00:00',
      apprenticeship_manager: 'Njenga Njunge',
      account_active: false,
      email_sent: false,
      project_id: 1,
      cohort: 'Lagos-43',
      project_role_name: 'Engineer',
      created_at: '2019-06-12 19:52:36',
      updated_at: '2019-06-12 19: 52: 36'
    },
    {
      id: 1,
      first_name: 'Steve',
      last_name: 'Madi',
      email: 'steve.madi@andela.com',
      picture:
        'https://media.haircutinspiration.com/photos/20181204023653/Back-and-Side-Square.jpg',
      sims_project: 'AuthorsHaven',
      sims_project_technology: 'Python/Django',
      sims_start_date: '2018-10-31 00:00:00',
      sims_end_date: '2019-01-10 00:00:00',
      sims_manager: 'Duncan Ireri',
      apprenticeship_project: 'WatchTower',
      apprenticeship_technology: 'Javascript/Ruby',
      apprenticeship_start_date: '2019-06-11 00:00:00',
      apprenticeship_end_date: '2019-06-26 00:00:00',
      apprenticeship_manager: 'Njenga Njunge',
      account_active: false,
      email_sent: false,
      project_id: 1,
      cohort: 'Lagos-42',
      project_role_name: 'FPC',
      created_at: '2019-06-12 19:52:36',
      updated_at: '2019-06-12 19: 52: 36'
    }
  ]
};
