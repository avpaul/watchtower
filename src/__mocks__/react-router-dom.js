import React from 'react';
import PropTypes from 'prop-types';

const rrd = require('react-router-dom');

// Just render plain div with its children
rrd.BrowserRouter = ({ children }) => <div>{children}</div>;

rrd.BrowserRouter.propTypes = {
  children: PropTypes.objectOf(PropTypes.element, PropTypes.func).isRequired,
};

module.exports = rrd;
