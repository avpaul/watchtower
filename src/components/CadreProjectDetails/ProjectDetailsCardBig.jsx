import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { formatText } from '../../services/helper';
import DocLogo from '../../static/doc.svg';

const ProjectDetailsCardBig = ({ projectDetails }) => {
  const { paragraph1, paragraph2 } = formatText(projectDetails.about);
  return (
    <Fragment>
      <div className="project-details__card-big">
        <h2>About {projectDetails.name}</h2>
        <p>{paragraph1}</p>

        <p>{paragraph2}</p>
        <h2>Relevant Documents</h2>
        <div className="relevant-documents">
          {projectDetails.documents &&
            projectDetails.documents.map(document => (
              <a href={document.url} download>
                <div className="relevant-documents-card">
                  <img src={DocLogo} alt="doc-logo" />
                  <span className="text-left w-75">
                    {document.name.substring(0, 20)}...
                  </span>
                </div>
              </a>
            ))}
        </div>
      </div>
    </Fragment>
  );
};
ProjectDetailsCardBig.propTypes = {
  projectDetails: PropTypes.shape({}).isRequired
};
export default ProjectDetailsCardBig;
