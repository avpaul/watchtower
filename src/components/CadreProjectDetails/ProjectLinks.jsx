import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import SlackLogo from '../../static/slack-logo-icon.png';
import InvisionLogo from '../../static/invision-logo.svg';

const ProjectLinks = ({ projectDetails }) => (
  <Fragment>
    <a
      href={
        projectDetails.channels.id
          ? `https://andela.slack.com/messages/CCU16SAKA/${projectDetails.channels.id}`
          : '#'
      }
    >
      <div className="links">
        <img src={SlackLogo} alt="slack-logo" className="mr-1" />
        {projectDetails.channels.id
          ? `#${projectDetails.name.substring(0, 10)}`
          : 'N/A'}
      </div>
    </a>
    <a
      href={projectDetails.mockups ? `${projectDetails.mockups}` : '#'}
      target="_blank"
      rel="noopener noreferrer"
    >
      <div className="links">
        <img src={InvisionLogo} alt="invision logo" className="mr-1" />
        {projectDetails.mockups
          ? `
        ${projectDetails.name}`
          : 'N/A'}
      </div>
    </a>
  </Fragment>
);
ProjectLinks.propTypes = {
  projectDetails: PropTypes.shape({}).isRequired
};
export default ProjectLinks;
