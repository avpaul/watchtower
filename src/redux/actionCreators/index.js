import test from '../constants';

export default () => ({
  type: test.TEST_REDUCER,
});
