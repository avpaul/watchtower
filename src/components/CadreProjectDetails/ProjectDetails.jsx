/* eslint-disable react/prefer-stateless-function */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ProjectDetailsCardBig from './ProjectDetailsCardBig';
import ProjectDetailsCardSmall from './ProjectDetailsCardSmall';
import ProjectEngineersTable from '../ProjectEngineersTable';
import './CadreProjectDetails.scss';
import ReturnButton from '../Buttons/ReturnButton';
import PMloader from '../CustomLoader/PMLoader';
import NotFoundPage from '../../views/NotFoundPage';

class ProjectDetails extends Component {
  constructor() {
    super();
    this.state = {
      projectDetails: null,
      roles: [],
      loading: true
    };
  }

  async componentDidMount() {
    const { match, getProject, fetchAllRoles } = this.props;
    let projectDetails;

    // Check if parameter in the browser is an Integer before making the API call to fetch the resource
    if (Number.isInteger(+match.params.id)) {
      await getProject(match.params.id);
      const { singleProject } = this.props;

      if (Object.keys(singleProject.data).length) {
        [projectDetails] = singleProject.data.project;
      }
    }

    await fetchAllRoles();
    const { cadreRoles: roles } = this.props;

    this.setState({ roles });
    this.setState({ projectDetails, loading: false });
  }

  render() {
    const { projectDetails, loading, roles } = this.state;
    const { history, match } = this.props;

    const projectBody = projectDetails ? (
      <div className="project-details">
        <div className="d-flex justify-content-between">
          <ReturnButton history={history} />
          <Link
            className="project-details__edit-btn"
            to={{
              pathname: `/cadre/projects/${parseInt(match.params.id, 10)}/edit`,
              projectDetails
            }}
          >
            {' '}
            EDIT PROJECT
          </Link>
        </div>
        <div className="project-details__main m-3">
          <ProjectDetailsCardBig projectDetails={projectDetails} />
          <ProjectDetailsCardSmall projectDetails={projectDetails} />
        </div>
        <div className="project-details__engineers">
          <h3>List of {projectDetails.name} engineers</h3>
          <ProjectEngineersTable
            engineers={projectDetails.engineers}
            roles={roles}
          />
        </div>
      </div>
    ) : (
      <NotFoundPage />
    );

    return loading ? <PMloader /> : projectBody;
  }
}

ProjectDetails.propTypes = {
  location: PropTypes.shape({}),
  history: PropTypes.shape({}).isRequired,
  match: PropTypes.shape({}).isRequired,
  getProject: PropTypes.func.isRequired,
  singleProject: PropTypes.shape().isRequired,
  cadreRoles: PropTypes.arrayOf(PropTypes.shape({})),
  fetchAllRoles: PropTypes.func.isRequired
};
ProjectDetails.defaultProps = {
  location: { state: 'something here' },
  cadreRoles: []
};
export default ProjectDetails;
