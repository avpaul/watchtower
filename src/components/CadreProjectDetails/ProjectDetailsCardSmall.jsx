import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import './CadreProjectDetails.scss';
import arrayKey from 'weak-key';
import DefaultManagerIcon from '../../static/profile.svg';
import DefaultProjectIcon from '../../static/projectIcon.png';
import ProjectLinks from './ProjectLinks';

const formatStacks = (technologies = []) =>
  technologies.map(technology => (
    <span className="stack" key={arrayKey(technology)}>
      {technology.name || ''}
    </span>
  ));
const ProjectDetailsCardSmall = ({
  projectDetails,
  projectDetails: { technologies }
}) => (
  <Fragment>
    <div className="project-details__card-small">
      <div className="project-details">
        <div className="project-type">
          <div>{projectDetails.type.toUpperCase()}</div>
        </div>
        <div className="project-logo">
          <img
            src={projectDetails.logo || DefaultProjectIcon}
            alt="project-logo"
          />
          <h3>{projectDetails.name}</h3>
        </div>
        <div className="project-theme">
          <p>{projectDetails.tagline}</p>
        </div>
        <div className="project-stacks">{formatStacks(technologies)}</div>
      </div>

      <div className="project-team">
        <h4>Project Manager</h4>
        <div className="project-manager-details">
          <img
            className="manager-photo"
            src={DefaultManagerIcon}
            alt="manager-pic"
          />
          <div className="manager-details">
            <h4>{projectDetails.manager.name}</h4>
            <p>{projectDetails.manager.email}</p>
          </div>
        </div>
      </div>

      <div className="project-links">
        <ProjectLinks projectDetails={projectDetails} />
      </div>
    </div>
  </Fragment>
);
ProjectDetailsCardSmall.propTypes = {
  projectDetails: PropTypes.shape({}).isRequired
};

export default ProjectDetailsCardSmall;
