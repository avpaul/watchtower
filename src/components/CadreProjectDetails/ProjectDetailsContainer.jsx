import { connect } from 'react-redux';
import ProjectDetails from './ProjectDetails';
import { getAProject } from '../../redux/actionCreators/projectsActions';

export const mapStateToProps = ({ allProjects, singleProject }) => ({
  allProjects,
  singleProject
});

const mapDispatchToProps = dispatch => ({
  getProject: id => dispatch(getAProject(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ProjectDetails);
