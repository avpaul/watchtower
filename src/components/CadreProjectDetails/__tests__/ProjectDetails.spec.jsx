import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from '../../../redux/actionCreators/projectsActions';
import ProjectDetails from '../ProjectDetails';
import ProjectDetailsCardBig from '../ProjectDetailsCardBig';
import ProjectDetailsCardSmall from '../ProjectDetailsCardSmall';
import ProjectDetailsContainer from '../ProjectDetailsContainer';
import projectDetails from '../../../__mocks__/projectDetails';
import ProjectLinks from '../ProjectLinks';

jest.spyOn(actions, 'getAProject');
actions.getAProject.mockImplementation(() => ({
  type: 'GET_SINGLE_PROJECT_REQUEST'
}));

describe('Project details card tests', () => {
  const props = {
    allProjects: {
      data: [projectDetails]
    },
    allRoles: {
      data: projectDetails.engineers.map(engineer => ({
        name: engineer.project_role_name,
        id: 1
      }))
    },
    match: {
      params: { id: 1 }
    },
    history: { replace: jest.fn() },
    getProject: {},
    getAProjectReducer: jest.fn()
  };

  const initialState = {
    singleProject: {
      data: {
        project: [projectDetails]
      }
    },
    allRoles: {
      data: projectDetails.engineers.map(engineer => ({
        name: engineer.project_role_name,
        id: 1
      }))
    },
    loading: true
  };

  const mockStore = configureStore([thunk]);
  const store = mockStore(initialState);

  const setup = (CardName, otherProps) => {
    const projectDetailsCardWrapper = shallow(
      <CardName {...props} projectDetails={otherProps} />
    );
    return projectDetailsCardWrapper;
  };

  it('should render project details page without crashing', () => {
    const { projectDetailsCardWrapper } = setup(ProjectDetails);
    expect(projectDetailsCardWrapper).toMatchSnapshot();
  });

  it('should render project details card BIG without crashing', () => {
    const { projectDetailsCardWrapper } = setup(
      ProjectDetailsCardBig,
      props.allProjects.data[0]
    );

    expect(projectDetailsCardWrapper).toMatchSnapshot();
  });
  it('should render project details card SMALL without crashing', () => {
    const { projectDetailsCardWrapper } = setup(
      ProjectDetailsCardSmall,
      props.allProjects.data[0]
    );
    expect(projectDetailsCardWrapper).toMatchSnapshot();
  });
  it('should render project links component successfully', () => {
    const projectLinkWrapper = shallow(
      <ProjectLinks projectDetails={projectDetails} />
    );
    expect(projectLinkWrapper).toMatchSnapshot();
  });

  it('should get a single project if no project details is available', () => {
    const newProps = {
      allProjects: {
        data: []
      },
      match: {
        params: { id: 1 }
      },
      history: { replace: jest.fn() },
      getProject: jest.fn()
    };

    mount(
      <MemoryRouter>
        <ProjectDetailsContainer store={store} {...newProps} />
      </MemoryRouter>
    );
    expect(store.getActions()[0].type).toBe('GET_SINGLE_PROJECT_REQUEST');
  });
});
