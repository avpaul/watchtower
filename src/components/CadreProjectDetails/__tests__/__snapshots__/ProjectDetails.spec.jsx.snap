// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project details card tests should render project details card BIG without crashing 1`] = `undefined`;

exports[`Project details card tests should render project details card SMALL without crashing 1`] = `undefined`;

exports[`Project details card tests should render project details page without crashing 1`] = `undefined`;

exports[`Project details card tests should render project links component successfully 1`] = `
<Fragment>
  <a
    href="https://andela.slack.com/messages/CCU16SAKA/undefined"
    rel="noopener noreferrer"
    target="_blank"
  >
    <div
      className="links"
    >
      <img
        alt="slack-logo"
        src="slack-logo-icon.png"
      />
      #
      test proje
      ... team on slack
    </div>
  </a>
  <a
    href="http://google.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    <div
      className="links"
    >
      <img
        alt="invision logo"
        src="invision-logo.svg"
      />
      test project
       on invision
    </div>
  </a>
</Fragment>
`;
